Documentation/Evolution/RegexSyntax.md

```diff
@@ -0,0 +1,245 @@
+
+We aim to parse a superset of the syntax accepted by a variety of popular regular expression engines.
+
+**TODO: Elaborate**
+
+## Engines supported
+
+We aim to implement a syntactic superset of:
+
+- [PCRE 2][pcre2-syntax], an "industry standard" of sorts, and a rough superset of Perl, Python, etc.
+- [Oniguruma][oniguruma-syntax], an internationalization-oriented engine with some modern features
+- [ICU][icu-syntax], used by NSRegularExpression, a Unicode-focused engine.
+- [.NET][.net-syntax]'s regular expressions, which support delimiter-balancing and some interesting minor details on conditional patterns.
+- **TODO: List Java here? It doesn't really add any more syntax than the above other than `\p{javaLowerCase}`**
+
+We also intend to achieve at least Level 1 (**TODO: do we want to promise Level 2?**) [UTS#18][uts18] conformance, which specifies regular expression matching semantics without mandating any particular syntax. However we can infer syntactic feature sets from its guidance.
+
+## Regex syntax supported
+
+### General syntax
+
+The following syntax are supported by all the above engines.
+
+#### Alternation
+
+```
+Regex       -> '' | Alternation
+Alternation -> Concatenation ('|' Concatenation)*
+```
+
+This is the operator with the lowest precedence in a regular expression, and checks if any of its branches match the input.
+
+#### Concatenation
+
+```
+Concatenation   -> (!'|' !')' ConcatComponent)*
+ConcatComponent -> Trivia | Quote | Quantification
+```
+
+Implicitly denoted by adjacent expressions, a concatenation matches against a sequence of regular expression patterns. This has a higher precedence than an alternation, so e.g `abc|def` matches against `abc` or `def`. The `ConcatComponent` token varies across engine, but at least matches some form of trivia, e.g comments, quoted sequences e.g `\Q...\E`, and a quantified expression.
+
+#### Quantification
+
+```
+Quantification -> QuantOperand Quantifier?
+Quantifier     -> ('*' | '+' | '?' | '{' Range '}') QuantKind?
+QuantKind      -> '?' | '+'
+```
+
+Specifies that the operand may be matched against a certain number of times.
+
+#### Groups
+
+```
+GroupStart    -> '(?' GroupKind | '('
+GroupKind     -> ':' | '|' | '>' | '=' | '!' | '*' | '<=' | '<!' | '<*'
+               | NamedGroup | MatchingOptionSeq (':' | ')')
+            
+NamedGroup    -> 'P<' GroupNameBody '>'
+               | '<' GroupNameBody '>'
+               | "'" GroupNameBody "'"
+
+GroupNameBody -> Identifier
+```
+
+Groups define a new scope within which a recursive regular expression pattern may occur. Groups have different semantics depending on how they are introduced, some may capture the nested match, some may match against the input without advancing, some may change the matching options set in the new scope, etc.
+
+#### Anchors
+
+```
+Anchor -> '^' | '$' | '\b'
+```
+
+Anchors match against a certain position in the input rather than on a particular character of the input.
```

Comment:
```suggestion
Anchors match against a certain position in the input rather than on a particular character of the input.

**TODO: We support a lot more anchors. Might be worth listing them with a very brief description of what they mean.**
```

 Need to balance this with brevity, of course. We can use disclosure triangles for anything that gets too in-depth, like options and nitty-gritty engine differences.